### ISSUE TITLE
Has anyone managed to reproduce the TimeSformer (the basic model) results on K400?

### ISSUE THREAD

**Issue by kami93** [facebookresearch/TimeSformer#94]
Hi. Thanks for this nice work.

I am trying to reproduce the basic (divST_8x32_224) model's results reported in this repository (acc@1 77.9 / acc@5 93.2) on my computer. I've downloaded K400 from the [cvfoundation's github repo](https://github.com/cvdfoundation/kinetics-dataset) provided at ([https://github.com/facebookresearch/TimeSformer/blob/main/timesformer/datasets/DATASET.md](https://github.com/facebookresearch/TimeSformer/blob/main/timesformer/datasets/DATASET.md))  and tried the following experiments/tests. Directly stating the results, I cannot reproduce the reported accuracies on the K400 dataset.  Please check the following details and logs. I would appreciate any comment please :).

### 1. General Information

- Model of Interest: the basic (divST_8x32_224) TimeSformer model (defined by `TimeSformer_divST_8x32_224_4gpus.yaml`). I used most settings except batch_size and num_workers unchanged from the default setting in this YAML file.

- GPU Setting: 4 X Nvidia 3090, which accepts 8 batches per GPU (thus 32 online batch size is used). However, please note that the default implementation in this repo performs gradient accumulation and the effective batch size is 64. ([https://github.com/facebookresearch/TimeSformer/blob/main/tools/train_net.py#L102-L116](https://github.com/facebookresearch/TimeSformer/blob/main/tools/train_net.py#L102-L116))

- Testing protocol
-- I used the `TimeSformer_divST_8x32_224_TEST.yaml`, which does `1 NUM_ENSEMBLE_VIEWS` and `3 NUM_SPATIAL_CROPS` for testing.

- DATASET quality information:
-- Applied the `replacement_for_corrupted_k400.tgz` provided by the [cvfoundation repo](https://github.com/cvdfoundation/kinetics-dataset)
-- Manually repaired some broken files
-- Down-scaled mp4 videos to 256p, using [ffmpeg-python](https://github.com/kkroening/ffmpeg-python) package with only using the "scale" filter, which produces an equivalent ffmpeg command of:
`ffmpeg -i video-big.mp4 -filter_complex "[0]scale=-2:256[s0]" -map "[s0]" video-256p.mp4 `
or
`ffmpeg -i video-big.mp4 -filter_complex "[0]scale=256:-2[s0]" -map "[s0]" video-256p.mp4 `
for horizontal videos and vertical videos, respectively.

- Statistics
-- Train set: 241,199 samples ( - 5,335 samples after removing corrupted files, compared to 246,534 samples originally annotated by the train.csv provided by cvfoundation repo)
-- Val set:  19,875 samples ( - 31 samples after removing corrupted files, compared to 19,906 samples originally annotated by the val.csv provided by cvfoundation repo).
_-- Test Set: 38,682 samples ( - 1,123 samples after removing corrupted files, compared to 39,805 samples originally annotated by the test.csv provided by cvfoundation repo)_
**IMPORTANT note: the validation set is used for evaluating models in this repo. The test set statistic is written just for the sake of providing complete information.** (Meanwhile, It would be informative if the authors can tell us why the _actual_ test set was not used for evaluating their models)

- Additional Comment
As noticed by many people here, some data samples' time spans are shorter than 10 seconds, which is the original time span designed by the kinetics-400 dataset's authors. There are mixed reasons for this issue (e.g., wrong label (the YouTube source video isn't as long as the annotated time duration, downloading error (YouTube sometimes detects that you are crawling their videos, and suddenly breaks the stream connection while downloading), etc.)
**Anyhow, this issue should commonly exist, for 100% same files for every one of us who downloaded kinetics-400 from [the cvfoundation repo](https://github.com/cvdfoundation/kinetics-dataset), not directly crawled it from YouTube. Hence, the reported accuracy should be reproduced regardless of the issue. I assume authors also have used the same version of kinetics-400 dataset from cvfoundation repo, as noted in [README.md](https://github.com/facebookresearch/TimeSformer/blob/main/timesformer/datasets/DATASET.md).**

### 2. Experiments

- 2-1. Directly testing the pretrained model parameter provided by the authors ([https://www.dropbox.com/s/g5t24we9gl5yk88/TimeSformer_divST_8x32_224_K400.pyth](https://www.dropbox.com/s/g5t24we9gl5yk88/TimeSformer_divST_8x32_224_K400.pyth))

- 2-2. Training from scratch (of course, starting from ViT-B ImageNet-21k weights) with kinetics-400 from cvfoundation repo
-- Used the default `TimeSformer_divST_8x32_224_4gpus.yaml` with altering TRAIN.BATCH_SIZE to 32 and DATA_LOADER.NUM_WORKERS to 16, which should result in the following important hyperparameters:
--- base lr: 0.005
--- lt_deacy at steps: [0, 11, 14]
--- lr_decay multipliers: [1, 0.1, 0.01]
--- Epochs: 15
--- Optimizer: SGD with 0.9 momentum and 1e-4 weight decays
--- batch_size (per iteration): 32
--- effective batch_size (due to the gradient accumulation): 64
--- num_frames: 8
--- sampling_rate: 32
...

| name | dataset | # of frames | spatial crop | acc@1 | acc@5 | 
| --- | --- | --- | --- | --- | --- | 
| TimeSformer (Reported) | K400 | 8 | 224 | 77.9 | 93.2 |
| TimeSformer (Experiment 2-1) | K400 | 8 | 224 | 77.38 (-0.52) | 93.16 (-0.04) |
| TimeSformer (Experiment 2-2) | K400 | 8 | 224 | 76.16 (-1.74)  | 92.43 (-0.77) |

### 3. Discussion
Overall, both Experiment 2-1 and Experiment 2-2 resulted in non-negligible (>0.5) acc@1 drops compared to the table reported in this repo. As can be seen from the attached full train/test logs, there have been no warnings and/or errors claimed by the dataloader (I carefully filtered out all the erroneous samples prior to experiments, which resulted in 241,199 train samples and 19,875 evaluation samples ).

One obvious thing that I am uncertain about the authors' experimental setting is precise parameters while converting videos to 256p resolution (e.g., bitrate, codecs, etc...), however, I guess the potential accuracy drop due to their difference should be subtle.

For Experiment 2-2 (training from scratch), I understand that there can be many different factors that can cause differences in the end-performance. However, for Experiment 2-1, it is really strange that the acc@1 is still not being matched even though I used the provided pre-trained weights, and the evaluation samples (the validation set) are almost the same as the list of files defined in the val.csv provided by cvfoundation repo (-31 difference in sample counts, after I filtered out corrupted files).

I would really appreciate any tips and comments about reproducing the results.

### 4. Attached Logs
[Testing log for experiment 2-1](https://drive.google.com/file/d/1vppqBrWRo9bBgW3WxRsuIRnYmoDyMn9w/view?usp=sharing)
[Training log for experiment 2-2](https://drive.google.com/file/d/1AdTLsUC2oxLU0AB5q1ENOpX8CS_ttuGL/view?usp=sharing)
[Testing log for experiment 2-2](https://drive.google.com/file/d/1qEasPa94cnNtlMZg5yha33aHaCXZzzVE/view?usp=sharing)

**Comment by gberta** [comment]
It's difficult to say why you are getting a drop in performance without the ability to debug myself. I thoroughly tested the implementation on a different number of GPUs, batch sizes, etc. before releasing the code, and I was able to reproduce results in these settings. Therefore, I would suggest to double check the data processing pipeline as that's the major difference between our two settings.

**Comment by JosephChenHub** [comment]
> Hi. Thanks for this nice work.
> 
> I am trying to reproduce the basic (divST_8x32_224) model's results reported in this repository (acc@1 77.9 / acc@5 93.2) on my computer. I've downloaded K400 from the [cvfoundation's github repo](https://github.com/cvdfoundation/kinetics-dataset) provided at (https://github.com/facebookresearch/TimeSformer/blob/main/timesformer/datasets/DATASET.md) and tried the following experiments/tests. Directly stating the results, I cannot reproduce the reported accuracies on the K400 dataset. Please check the following details and logs. I would appreciate any comment please :).
> 
> ### 1. General Information
> * Model of Interest: the basic (divST_8x32_224) TimeSformer model (defined by `TimeSformer_divST_8x32_224_4gpus.yaml`). I used most settings except batch_size and num_workers unchanged from the default setting in this YAML file.
> * GPU Setting: 4 X Nvidia 3090, which accepts 8 batches per GPU (thus 32 online batch size is used). However, please note that the default implementation in this repo performs gradient accumulation and the effective batch size is 64. (https://github.com/facebookresearch/TimeSformer/blob/main/tools/train_net.py#L102-L116)
> * Testing protocol
>   -- I used the `TimeSformer_divST_8x32_224_TEST.yaml`, which does `1 NUM_ENSEMBLE_VIEWS` and `3 NUM_SPATIAL_CROPS` for testing.
> * DATASET quality information:
>   -- Applied the `replacement_for_corrupted_k400.tgz` provided by the [cvfoundation repo](https://github.com/cvdfoundation/kinetics-dataset)
>   -- Manually repaired some broken files
>   -- Down-scaled mp4 videos to 256p, using [ffmpeg-python](https://github.com/kkroening/ffmpeg-python) package with only using the "scale" filter, which produces an equivalent ffmpeg command of:
>   `ffmpeg -i video-big.mp4 -filter_complex "[0]scale=-2:256[s0]" -map "[s0]" video-256p.mp4 `
>   or
>   `ffmpeg -i video-big.mp4 -filter_complex "[0]scale=256:-2[s0]" -map "[s0]" video-256p.mp4 `
>   for horizontal videos and vertical videos, respectively.
> * Statistics
>   -- Train set: 241,199 samples ( - 5,335 samples after removing corrupted files, compared to 246,534 samples originally annotated by the train.csv provided by cvfoundation repo)
>   -- Val set:  19,875 samples ( - 31 samples after removing corrupted files, compared to 19,906 samples originally annotated by the val.csv provided by cvfoundation repo).
>   _-- Test Set: 38,682 samples ( - 1,123 samples after removing corrupted files, compared to 39,805 samples originally annotated by the test.csv provided by cvfoundation repo)_
>   **IMPORTANT note: the validation set is used for evaluating models in this repo. The test set statistic is written just for the sake of providing complete information.** (Meanwhile, It would be informative if the authors can tell us why the _actual_ test set was not used for evaluating their models)
> * Additional Comment
>   As noticed by many people here, some data samples' time spans are shorter than 10 seconds, which is the original time span designed by the kinetics-400 dataset's authors. There are mixed reasons for this issue (e.g., wrong label (the YouTube source video isn't as long as the annotated time duration, downloading error (YouTube sometimes detects that you are crawling their videos, and suddenly breaks the stream connection while downloading), etc.)
>   **Anyhow, this issue should commonly exist, for 100% same files for every one of us who downloaded kinetics-400 from [the cvfoundation repo](https://github.com/cvdfoundation/kinetics-dataset), not directly crawled it from YouTube. Hence, the reported accuracy should be reproduced regardless of the issue. I assume authors also have used the same version of kinetics-400 dataset from cvfoundation repo, as noted in [README.md](https://github.com/facebookresearch/TimeSformer/blob/main/timesformer/datasets/DATASET.md).**
> 
> ### 2. Experiments
> * 2-1. Directly testing the pretrained model parameter provided by the authors (https://www.dropbox.com/s/g5t24we9gl5yk88/TimeSformer_divST_8x32_224_K400.pyth)
> * 2-2. Training from scratch (of course, starting from ViT-B ImageNet-21k weights) with kinetics-400 from cvfoundation repo
>   -- Used the default `TimeSformer_divST_8x32_224_4gpus.yaml` with altering TRAIN.BATCH_SIZE to 32 and DATA_LOADER.NUM_WORKERS to 16, which should result in the following important hyperparameters:
>   --- base lr: 0.005
>   --- lt_deacy at steps: [0, 11, 14]
>   --- lr_decay multipliers: [1, 0.1, 0.01]
>   --- Epochs: 15
>   --- Optimizer: SGD with 0.9 momentum and 1e-4 weight decays
>   --- batch_size (per iteration): 32
>   --- effective batch_size (due to the gradient accumulation): 64
>   --- num_frames: 8
>   --- sampling_rate: 32
>   ...
> 
> name	dataset	# of frames	spatial crop	acc@1	acc@5
> TimeSformer (Reported)	K400	8	224	77.9	93.2
> TimeSformer (Experiment 2-1)	K400	8	224	77.38 (-0.52)	93.16 (-0.04)
> TimeSformer (Experiment 2-2)	K400	8	224	76.16 (-1.74)	92.43 (-0.77)
> ### 3. Discussion
> Overall, both Experiment 2-1 and Experiment 2-2 resulted in non-negligible (>0.5) acc@1 drops compared to the table reported in this repo. As can be seen from the attached full train/test logs, there have been no warnings and/or errors claimed by the dataloader (I carefully filtered out all the erroneous samples prior to experiments, which resulted in 241,199 train samples and 19,875 evaluation samples ).
> 
> One obvious thing that I am uncertain about the authors' experimental setting is precise parameters while converting videos to 256p resolution (e.g., bitrate, codecs, etc...), however, I guess the potential accuracy drop due to their difference should be subtle.
> 
> For Experiment 2-2 (training from scratch), I understand that there can be many different factors that can cause differences in the end-performance. However, for Experiment 2-1, it is really strange that the acc@1 is still not being matched even though I used the provided pre-trained weights, and the evaluation samples (the validation set) are almost the same as the list of files defined in the val.csv provided by cvfoundation repo (-31 difference in sample counts, after I filtered out corrupted files).
> 
> I would really appreciate any tips and comments about reproducing the results.
> 
> ### 4. Attached Logs
> [Testing log for experiment 2-1](https://drive.google.com/file/d/1vppqBrWRo9bBgW3WxRsuIRnYmoDyMn9w/view?usp=sharing) [Training log for experiment 2-2](https://drive.google.com/file/d/1AdTLsUC2oxLU0AB5q1ENOpX8CS_ttuGL/view?usp=sharing) [Testing log for experiment 2-2](https://drive.google.com/file/d/1qEasPa94cnNtlMZg5yha33aHaCXZzzVE/view?usp=sharing)

Hi @gberta , I observe the similar case. For directly testing 'TimeSformer_divST_8x32_224_K400.pyth' on validation set of kinetics 400 (19,877 samples after removing corrupted videos),  the result is 
<!DOCTYPE html>

name | dataset | # of frames | spatial crop | acc@1 | acc@5
-- | -- | -- | -- | -- | --
TimeSformer (Reported) | K400 | 8 | 224 | 77.9 | 93.2
TimeSformer (test) | K400 | 8 | 224 | 77.59 | 93.22
